modules/login/php/passwordreset.class.inc

```diff
@@ -73,45 +74,78 @@ class PasswordReset extends \NDB_Form
      *
      * @param array $values The values that were submitted to the page.
      *
-     * @return none
+     * @return void
      */
     function _process($values)
     {
+        $DB         =& \Database::singleton();
+        $resetLimit =10;
+
         // Always claim success to phishers.
         $this->tpl_data['success'] = 'Password reset. '
-        . 'You should receive an email within a few minutes.';
+            . 'You should receive an email within a few minutes.';
+
+        // Check how many reset requests this IP has made
+        $userIP       = ip2long($_SERVER['REMOTE_ADDR']);
+        $numberResets = intval(
+            $DB->pselectOne(
+                "SELECT COUNT(*) FROM password_recovery WHERE ip=:ip",
+                array('ip' => $userIP)
+            )
+        );
 
         // create the user object
         $username = $values['username'];
         $user     =& \User::singleton($username);
 
-        $email = $user->getData('Email');
-
-        // check that it is a valid user
-        if (empty($email) || empty($user)) {
+        if (empty($user)) {
+            // the user does not exist in the databse,
+            // pretend like everything is fine
             return;
         }
+        if ($numberResets>$resetLimit) {
+            //Do something or nothing
+            return;
+        }
+
+        $client = new \NDB_Client();
+        $client->makeCommandLine();
+
+        $email  = $user->getData('Email');
+        $name   = $user->getFullname();
+        $userId = $user->getId();
+
+        $length = 4;
+        $token  = bin2hex(random_bytes($length));
+
+        // TODO: problem, the insert below generate SQL FK errors because there
+        // in no logged in user to track history. see database.class.inc line 895
+        // user is being set to 'unknown' at this stage
+        $DB->insert(
```

Comment:
@driusan need to resolve issue above

Reply:
The history table shouldn't have any foreign keys. It should be denormalized because it's a history table.. I think the solution is just to remove the constraint.

Reply:
okai so for now can I just ignore the warnings ?

Reply:
yeah, I think so.